pkg/destinations/redshift/insert.go

```diff
@@ -0,0 +1,170 @@
+package redshift
+
+import (
+	"path/filepath"
+
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/rs/zerolog/log"
+	"github.com/scratchdata/scratchdata/util"
+
+	"github.com/scratchdata/scratchdata/pkg/storage/blobstore/s3"
+)
+
+func (s *RedshiftServer) getGolumnNames(table string) (map[string]bool, error) {
+	schema := "public"
+	if s.Schema != "" {
+		schema = s.Schema
+	}
+
+	sql := `
+		select "column" as column_name
+		from pg_table_def
+		where schemaname = $1 and tablename = $2
+	`
+
+	m := map[string]bool{}
+	rows, err := s.conn.Query(sql, schema, table)
+	if err != nil {
+		return nil, fmt.Errorf("getGolumnNames: cannot fetch column names: %w", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+
+		name := ""
+		if err := rows.Scan(&name); err != nil {
+			return nil, fmt.Errorf("getGolumnNames: cannot scan column name: %w", err)
+		}
+		m[strings.ToLower(name)] = true
+	}
+	if len(m) == 0 {
+		return nil, fmt.Errorf("getGolumnNames: no columns found: %w", err)
+	}
+	return m, nil
+}
+
+func (s *RedshiftServer) createColumns(table string, jsonTypes map[string]string) error {
+	cols, err := s.getGolumnNames(table)
+	if err != nil {
+		return err
+	}
+
+	log.Printf("Found existing columns %v", cols)
+
+	for colName, jsonType := range jsonTypes {
+		colType := "VARCHAR"
+		switch jsonType {
+		case "int":
+			colType = "BIGINT"
+		case "bool":
+			colType = "BOOLEAN"
+		case "float":
+			colType = "DOUBLE PRECISION"
+		case "string":
+			colType = "VARCHAR"
+		default:
+			colType = "VARCHAR"
+		}
+
+		if _, ok := cols[strings.ToLower(colName)]; !ok {
+			log.Printf("Column %s does not exist, creating it", colName)
+			sql := fmt.Sprintf("ALTER TABLE \"%s\" ADD COLUMN \"%s\" %s", table, colName, colType)
+			_, err = s.conn.Exec(sql)
```

Comment:
Can we check the type of error here? If two processes attempt to insert data at the same time, then we might find that some of the columns have already been created even when we execute this.

If the error is that the column already exists, then we can ignore the error and continue inserting. However if there is a different error, then we need to stop the insert process.

Reply:
So what are suggesting is too directly executing the Alter Add Column command?
Then if error is about already column existing, move ahead, else throw error?

Reply:
Correct